README.md

```diff
@@ -9,15 +23,67 @@ Basics
   standard Python naming conventions of being in snake\_case\_naming\_schemes.
 * Names should be mnemonic and descriptive and not strive to shorten more than necessary. Systems
   support long identifier names, so use them to be descriptive
 * All defaults and all arguments to a role should have a name that begins with the role name to help
   avoid collision with other names. Avoid names like `packages` in favor of a name like `foo_packages`.
+  (Rationale: Ansible has no namespaces, doing so reduces the potential for conflicts and makes
+  clear what role a given variable belongs to.)
+* Same argument applies for modules provided in the roles, they also need a `$ROLENAME_` prefix:
+  `foo_module`. While they are usually implementation details and not intended for direct use in
+  playbooks, the unfortunate fact is that importing a role makes them available to the rest of the
+  playbook and therefore creates opportunities for name collisions.
+* Moreover, internal variables (those that are not expected to be set by users) are to be prefixed
+  by two underscores: `__foo_variable`. (Rationale: role variables, registered variables, custom
+  facts are usually intended to be local to the role, but in reality are not local to the role - as
+  such a concept does not exist, and pollute the global namespace. Using the name of the role
+  reduces the potential for name conflicts and using the underscores clearly marks the variables as
+  internals and not part of the common interface. The two underscores convention has prior art in
+  some popular roles like
+  [geerlingguy.ansible-role-apache](https://github.com/geerlingguy/ansible-role-apache/blob/f2b91ac84001db3fd4b43306a8f73f1a54f96f7d/vars/Debian.yml#L8)). This
+  includes variables set by set_fact and register, because they persist in the namespace after the
+  role has finished!
+* Do not use special characters other than underscore in variable names, even if YAML/JSON allow
+  them. (Using such variables in Jinja2 or Python would be then very confusing and probably not
+  functional.)
+  
+## Providers
+
+When there are multiple implementations of the same functionality, we call them “providers”. A role
+supporting multiple providers should have an input variable called `$ROLENAME_provider`. If this
+variable is not defined, the role should detect the currently running provider on the system, and
+respect it. (Rationale: users can be surprised if the role changes the provider if they are running
+one already.) If there is no provider currently running, the role should select one according to the
+OS version. (E.g. on RHEL 7, chrony should be selected as the provider of time synchronization,
+unless there is ntpd already running on the system, or user requests it specifically. Chrony should
+be chosen on RHEL 8 as well, because it is the only provider available.) The role should set a
+variable or custom fact called `$ROLENAME_provider_os_default` to the appropriate default value for
```

Comment:
What if a role has multiple providers (e.g. one to determine the webserver daemon, and another to determine the JVM)? What's the naming scheme, then?

Reply:
We have not discussed this, but the timesync role provides an example: the provider variable for the ntp functionality is called `timesync_ntp_provider`, leaving open the possibility of adding `timesync_somethingelse_provider`: https://github.com/linux-system-roles/timesync/blob/f1a23686dc1107520a18b41c5d1fc52ee60d5fc5/defaults/main.yml#L7